File: solutions/2343-count-unguarded-cells-in-the-grid/count-unguarded-cells-in-the-grid.js
var countUnguarded = function(m, n, guards, walls) {
    // Initialize grid with zeros
    const g = Array(m).fill().map(() => Array(n).fill(0));
    
    // Mark guards and walls as 2
    for (const [x, y] of guards) {
        g[x][y] = 2;
    }
    for (const [x, y] of walls) {
        g[x][y] = 2;
    }
    
    // Directions: up, right, down, left
    const dirs = [[-1, 0], [0, 1], [1, 0], [0, -1]];
    
    // Process each guard's line of sight
    for (const [gx, gy] of guards) {
        for (const [dx, dy] of dirs) {
            let x = gx;
            let y = gy;
            while (true) {
                x += dx;
                y += dy;
                // Check cells in current direction until hitting boundary or obstacle
                if (x < 0 || x >= m || y < 0 || y >= n || g[x][y] === 2) {
                    break;
                }
                g[x][y] = 1;
            }
        }
    }
    
    // Count unguarded cells (cells with value 0)
    return g.reduce((sum, row) => sum + row.filter(cell => cell === 0).length, 0);
};

File: solutions/2343-count-unguarded-cells-in-the-grid/count-unguarded-cells-in-the-grid.swift
class Solution {
    func countUnguarded(_ m: Int, _ n: Int, _ guards: [[Int]], _ walls: [[Int]]) -> Int {
        // Initialize grid with zeros
        var g = Array(repeating: Array(repeating: 0, count: n), count: m)
        
        // Mark guards and walls as 2
        for e in guards {
            g[e[0]][e[1]] = 2
        }
        for e in walls {
            g[e[0]][e[1]] = 2
        }
        
        // Directions: up, right, down, left
        let dirs = [-1, 0, 1, 0, -1]
        
        // Process each guard's line of sight
        for e in guards {
            for k in 0..<4 {
                var x = e[0], y = e[1]
                let dx = dirs[k], dy = dirs[k + 1]
                
                // Check cells in current direction until hitting boundary or obstacle
                while x + dx >= 0 && x + dx < m && y + dy >= 0 && y + dy < n && g[x + dx][y + dy] < 2 {
                    x += dx
                    y += dy
                    g[x][y] = 1
                }
            }
        }
        
        // Count unguarded cells (cells with value 0)
        return g.reduce(0) { $0 + $1.filter { $0 == 0 }.count }
    }
}